### PR TITLE
Merge in importer.

### PR DESCRIPTION
**Pull request by DArtagan**
When duplicate records are found during import, the user has the option to merge the two sets together.  Presently, only works when tracks don't conflict.

This is not the final commit, I'm planning on resolving track conflicts before this is ready for merging.  Before I head down that road though, I'd really like my work double-checked (@sampsyo).  This project is impressively large, I did my best to understand this portion, but I'm not sure if I'm abusing the database or pipeline or something.  

Cheers!

P.S. Related to issue #112

### CODE REVIEW

**Comment by DArtagan** [comment]
@sampsyo a few questions:
* In for the unittests, I wanted to test merging against an album with multiple tracks.  Making a list of [a couple TrackInfo objects](https://github.com/sampsyo/beets/pull/1535/files#diff-e913c0058cb64594f017576823c46c3dR1159) only seems to ever have the first one picked up, why?
* I'd like to also allow for interactive merging, where two albums contain overlapping parts of each other.  My first thought is to compare the track ids, would you agree with this approach?  Would mocks be the proper way to unittest such interactions?
* I'm not sure why the builds are failing on test.test_mb.MBLibraryTest.test_match_album.  It built fine when I committed last weekend, but not this weekend?  All 37 tests in test_mb.py pass on my machine.
* I see you're running AppVeyor in addition to Travis CI.  What do you think of it?

**Comment by sampsyo** [comment]
Hi! Thank you for writing this! It looks amazing, and it fulfills something that many people have asked for. And I'm *very very* sorry that I didn't get around to this earlier—thanks for your patience while I caught up on the backlog.

I'll review the code now. To answer your questions:

> In for the unittests, I wanted to test merging against an album with multiple tracks. Making a list of a couple TrackInfo objects only seems to ever have the first one picked up, why?

I'm not 100% sure what's going on here—are you referring to the new `test_merge_duplicate_album_without_duplicate_tracks`, but with the commented-out lines commented back in? If so, should I just run that to see what you're talking about?

> I'd like to also allow for interactive merging, where two albums contain overlapping parts of each other. My first thought is to compare the track ids, would you agree with this approach?

Is the idea to let you *break apart* an album and then merge *part* of it into a new album? That sounds like a rather complex thing to orchestrate. Would you mind terribly much if we left that extension until after we finish this more basic version?

> Would mocks be the proper way to unittest such interactions?

Probably—depending on where the user input comes in.

> I'm not sure why the builds are failing on test.test_mb.MBLibraryTest.test_match_album. It built fine when I committed last weekend, but not this weekend? All 37 tests in test_mb.py pass on my machine.

That's not your fault, and now fixed. Nothing to worry about!

> I see you're running AppVeyor in addition to Travis CI. What do you think of it?

So far so good; it seems to do its job admirably, and it's great that it's free for open source. It's disabled for us, unfortunately, because we haven't gotten all the tests passing on Windows quite yet (#670).

File: beets/ui/commands.py [hunk 1]
```diff
@@ -810,6 +810,10 @@ def resolve_duplicate(self, task, found_duplicates):
         elif sel == 'k':
             # Keep both. Do nothing; leave the choice intact.
             pass
+        elif sel == 'm':
+            # Merge the two together
```
Comment (sampsyo) [inline beets/ui/commands.py]: Style nitpick: for comments like this, please use a period at the end of sentences.

File: beets/ui/commands.py [hunk 2]
```diff
@@ -810,6 +810,10 @@ def resolve_duplicate(self, task, found_duplicates):
...
+            task.should_merge_duplicates = True
+            print('should merge duplicates')
```
Comment (sampsyo) [inline beets/ui/commands.py]: Stray debug print. :smiley:

File: beets/importer.py [hunk 3]
```diff
@@ -520,6 +523,34 @@ def remove_duplicates(self, lib):
...
+        # choice_flag used for self.import_items() in other methods
+        self.choice_flag = action.ASIS
+
+        for item in existing_album.items():
+            self.items.append(item)
+        self.align_album_level_fields()
+
+        # Set ids to None so add() "add"s rather than "store"s in lib.add_album
+        for item in self.items:
+            item.id = None
```
Comment (sampsyo) [inline beets/importer.py]: Can you explain a little bit more about why this is necessary? Why should we re-add items that are already in the library, just associated with a different album?

File: beets/importer.py [hunk 4]
```diff
@@ -520,6 +523,34 @@ def remove_duplicates(self, lib):
...
+        """
+
+        # Tag all with the same album id
+        existing_album = self.find_duplicates(lib)[0]
+        for item in self.items:
+            item.album_id = existing_album.id
+
+        # Must set album_id before setting choice_flag
+        # choice_flag used for self.import_items() in other methods
+        self.choice_flag = action.ASIS
```
Comment (sampsyo) [inline beets/importer.py]: To check my understanding: It looks like this version of the feature doesn't *re-tag* the new, combined album. Instead, the match for the old album is ignored, and the files keep their current metadata. Is that right?

We may want to explore the other thing, which sounds like what people were originally interested in: the auto-tagger gets to "try again" with the new, larger album.

**Comment by DArtagan** [comment]
Mad props to udibox1209, making this happen in #2725.

**Pull request closed without merging**